Configurable deadlock detector interval for ingester. (resubmit)

marqc opened the pull request:
Resubmit because of github failures today, and previously submitter PR seems broken

## Which problem is this PR solving?
- Resolves #1125


## Short description of the changes
- Additional configurable flag for ingester deadlock detector interval
- Value of 0 disables deadlock_detector

yurishkuro commented on cmd/ingester/app/consumer/deadlock_detector.go:
```diff
@@ -138,21 +142,25 @@ func (s *deadlockDetector) start() {
-				if atomic.LoadUint64(detector.msgConsumed) == 0 {
-					s.panicFunc(-1)
-					return // For tests
+		if s.interval == 0 {
```
move to head of start() & exit immediately instead?

marqc replied:
Cannot really do that, because without initialization of "allPartitionsDeadlockDetector" API gets broken - and NullPointerExceptions will start to appear.

I little change this code not to check for interval inside goroutine.

yurishkuro commented on cmd/ingester/app/consumer/deadlock_detector_test.go:
```diff
@@ -112,3 +112,34 @@ func TestGlobalPanic(t *testing.T) {
+		},
+	}
+
+	d.start()
```
should include a call to close(), to ensure it works as expected when goroutines are not running

marqc replied:
added calls to close() and fixed blocked channels w/o goroutines with setting "closed" flag on detectors

yurishkuro commented on cmd/ingester/app/flags.go:
```diff
@@ -85,6 +91,10 @@ func AddFlags(flagSet *flag.FlagSet) {
+	flagSet.String(
+		ConfigPrefix+SuffixDeadlockInterval,
+		DefaultDeadlockInterval.String(),
+		"Interval to check for deadlocks. If no messages gets processed in given time, ingester app will exit. Value of 0 disable deadlock check.")
```
s/disable/disables/

marqc commented:
@yurishkuro I got rid of closed flag, and introduced disabled flag with short circuits as You suggested. Also changed logic to not create unnecessary channels.

yurishkuro commented on cmd/ingester/app/consumer/deadlock_detector.go:
```diff
@@ -82,18 +84,23 @@ func newDeadlockDetector(metricsFactory metrics.Factory, logger *zap.Logger, int
 	}
 
-	go s.monitorForPartition(w, partition)
+	if s.interval == 0 {
```
nit: w.disabled

yurishkuro commented on cmd/ingester/app/consumer/deadlock_detector.go:
```diff
@@ -82,18 +84,23 @@ func newDeadlockDetector(metricsFactory metrics.Factory, logger *zap.Logger, int
+		s.logger.Debug("Partition deadlock detector disabled")
+	} else {
+		w.closePartition = make(chan struct{}, 1)
+		w.done = make(chan struct{})
```
nit: I don't think it's necessary to pull these out of the main construction. It saves a couple of memory allocations, but makes the code harder to maintain.

yurishkuro commented on cmd/ingester/app/consumer/deadlock_detector.go:
```diff
@@ -133,52 +140,72 @@ func (s *deadlockDetector) start() {
 	}
 
-	go func() {
+	if s.interval == 0 {
```
detector.disabled

yurishkuro commented on cmd/ingester/app/consumer/deadlock_detector.go:
```diff
@@ -133,52 +140,72 @@ func (s *deadlockDetector) start() {
-				if atomic.LoadUint64(detector.msgConsumed) == 0 {
-					s.panicFunc(-1)
-					return // For tests
+		detector.done = make(chan struct{})
```
nit: don't pull this out, keep where it was

yurishkuro commented on cmd/ingester/app/consumer/deadlock_detector.go:
```diff
@@ -133,52 +140,72 @@ func (s *deadlockDetector) start() {
 }
 
 func (s *allPartitionsDeadlockDetector) incrementMsgCount() {
+	if s.disabled {
```
not necessary.

yurishkuro commented on cmd/ingester/app/consumer/deadlock_detector.go:
```diff
@@ -133,52 +140,72 @@ func (s *deadlockDetector) start() {
 }
 
 func (w *partitionDeadlockDetector) closePartitionChannel() chan struct{} {
+	if w.disabled {
```
not necessary

yurishkuro commented on cmd/ingester/app/consumer/deadlock_detector.go:
```diff
@@ -133,52 +140,72 @@ func (s *deadlockDetector) start() {
 }
 
 func (w *partitionDeadlockDetector) incrementMsgCount() {
+	if w.disabled {
```
not necessary

yurishkuro commented on cmd/ingester/app/flags_test.go:
```diff
@@ -51,4 +54,15 @@ func TestFlagDefaults(t *testing.T) {
+	assert.Equal(t, DefaultDeadlockInterval, o.DeadlockInterval)
+}
+
+func TestUnparsableDeadlockIntervalFlag(t *testing.T) {
```
Shouldn't it cause an error in flags parsing instead? This is what happens when I pass `hello` as a timeout
```
$ go run ./cmd/query/main.go --cassandra.timeout=hello
Error: invalid argument "hello" for "--cassandra.timeout" flag: time: invalid duration hello
Usage:
  jaeger-query [flags]
...
```

yurishkuro commented on cmd/ingester/app/consumer/deadlock_detector_test.go:
```diff
@@ -112,3 +112,54 @@ func TestGlobalPanic(t *testing.T) {
+	w.incrementMsgCount()
+	w.incrementAllPartitionMsgCount()
+	assert.Zero(t, len(w.closePartitionChannel()))
+	w.close()
```
I am not clear what this is testing. Why calling increments should have any effect on the closePartitionChannel?
